docs/architecture/architecture.md

```diff
@@ -0,0 +1,378 @@
+the virtual machine via a 9pfs virtio mount point. Hyperstart uses this mount point as the root
+filesystem for the container processes.
+
+![Overall architecture](overall-architecture.png)
+
+#### Hypervisor
+
+Clear Containers use [KVM](http://www.linux-kvm.org/page/Main_Page)/[QEMU](http://www.qemu-project.org/) to
+create virtual machines where Docker containers will run:
+
+![QEMU/KVM](qemu.png)
+
+** fixme - discuss different QEMUs - lite, q35, pc etc. **
+Although Clear Containers can run with any recent QEMU release, containers boot time and memory
+footprint are significantly optimized by using a specific QEMU version called [`qemu-lite`](https://github.com/clearcontainers/qemu/tree/qemu-lite-v2.9.0).
+
+`qemu-lite` improvements comes through a new `pc-lite` machine type, mostly by:
+- Removing many of the legacy hardware devices support so that the guest kernel does not waste
+time initializing devices of no use for containers.
+- Skipping the guest BIOS/firmware and jumping straight to the Clear Containers kernel.
+
+#### Agent
+
+['agent'](https://github.com/clearcontainers/agent) is a daemon running on the guest as an
+agent for managing containers and processes potentially running within those containers.
+
+The `agent` execution unit is the pod. An `agent` pod is a container sandbox defined
+by a set of namespaces (UTS, PID, mount and IPC). Although a pod can hold several containers,
+`cc-runtime` always runs a single container per pod. **fixme** incorrect<--
+
+**todo** add details on the agent protocol
+
+#### Runtime
+
+`cc-runtime` is an OCI compatible container runtime and is responsible for handling all
+commands specified by [the OCI runtime specification](https://github.com/opencontainers/runtime-spec)
+and launching `cc-shim` instances.
+
+Here we will describe how `cc-runtime` handles the most important OCI commands.
+
+##### `create`
+
+When handling the OCI `create` command, `cc-runtime` goes through the following steps:
+
+1. Create the container namespaces (Only the network and mount namespaces are currently supported),
+according to the container OCI configuration file.
+2. Spawn the `cc-shim` process and have it wait on a couple of temporary pipes for:
+   * A `cc-proxy` created file descriptor (one end of a socketpair) for the shim to connect to.
+   * The container `agent` sequence numbers for at most 2 I/O streams (One for `stdout` and `stdin`, one for `stderr`).
+   The `agent` uses those sequence numbers to multiplex all streams for all processes through one serial interface (The
+   virtio I/O serial one).
+3. Run all the [OCI hooks](https://github.com/opencontainers/runtime-spec/blob/master/config.md#hooks) in the container namespaces,
+as described by the OCI container configuration file.
+4. **fixme** [Set up the container networking](https://github.com/clearcontainers/runtime/blob/master/documentation/architecture.md#networking).
```

Comment:
fixme is just for the link location, right @grahamwhaley ?

Reply:
Looks like it. I believe we can change it to a 'local reference', so just `#networking` should resolve to the section in this document.

Reply:
done